File: js/jsfx.ts
// Copyright (c) 2022, the dartfx project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import './fx.js';

//declare function;
declare var jsfx: Function;
declare var jsfxWithEnvs: Function;
declare var jsfxAssignment: Function;
declare var jsSetFunctionResolver: Function;

/*
* Run expression and return the result.
*
* onGetEnvValue: Enviroment value callback function.
*/
export function fx(expression: string, onGetEnvValue?: Function): any {
	return jsfx(expression, onGetEnvValue);
}

/* 
* Run expression and return the result.
*
* envValues: Enviroment values set. If the expression contains a variable `$a.b$`. Then
*       you should give a values set like `{"a": {"b": something}}`.
*/
export function fxWithEnvs(expression: string, envValues: object): any {
	return jsfxWithEnvs(expression, envValues);
}

/* 
* Run assignment expression and return the right side value.
*
* expression: Assignment expression，eg: `$a.b$=1+2+3`.
* envValues: Enviroment values set. If the expression contains a variable `$a.b$`. Then
*       you should give a values set like `{"a": {"b": something}}`.
* leftEnvs: Used for [jsfxAssignment] in `jsfx.dart`.
*/
export function fxAssignment(expression: string, envValues: object): any {
	return jsfxAssignment(expression, envValues);
}

/*
* If there is a function witch the internal functions doesn't have. You can custom
* your own function by set a [FunctionResolver].
*/
export function fxSetFunctionResolver(functionResolver: Function) {
	jsSetFunctionResolver(functionResolver);
}

File: lib/src/jsfx.dart
import 'dart:js';

import 'dartfx_main.dart';
import 'util/jsvalue2dart.dart';

void main() {
  context['jsfx'] = jsfx;
  context['jsfxWithEnvs'] = jsfxWithEnvs;
  context['jsfxAssignment'] = jsfxAssignment;
  context['jsSetFunctionResolver'] = jsSetFunctionResolver;
}

dynamic jsSetFunctionResolver(JsFunction jsFunction) {
  fxSetFunctionResolver((name, arguments) {
    return jsFunction.apply([name, ...arguments]);
  });
}

dynamic jsfx(String expression, [JsFunction? onGetEnvValue]) {
  return fx(expression,
      onGetEnvValue: onGetEnvValue == null
          ? null
          : (envVar) {
              return onGetEnvValue.apply([envVar]);
            });
}

dynamic jsfxWithEnvs(String expression, JsObject envValues) {
  var envsMap = jsValueToDart(envValues);
  if (envsMap is Map) {
    return fxWithEnvs(expression, envsMap);
  } else {
    return null;
  }
}

dynamic jsfxAssignment(String expression, JsObject envValues) {
  var envsMap = jsValueToDart(envValues);

  if (envsMap is Map) {
    var obj = envValues;
    var valueKey = "";
    var rightValue = fxAssignment(expression, envsMap, leftEnvs: (envFields) {
      for (var i = 0; i < envFields.length - 1; i++) {
        obj = obj[envFields[i]];
      }

      valueKey = envFields.last;
    });
    obj[valueKey] = rightValue;
    return rightValue;
  }
}
